examples/stm32/f4/stm32f429i-discovery/lcd-dma/console.c

```diff
@@ -0,0 +1,274 @@
+ * support for editing characters (back space and delete)
+ * end when a <CR> character is received.
+ */
+int console_gets(char *s, int len)
+{
+	char *t = s;
+	char c;
+
+	*t = '\000';
+	/* read until a <CR> is received */
+	while ((c = console_getc(1)) != '\r') {
+		if ((c == '\010') || (c == '\127')) {
+			if (t > s) {
+				/* send ^H ^H to erase previous character */
+				console_puts("\010 \010");
+				t--;
+			}
+		} else {
+			*t = c;
+			console_putc(c);
+			if ((t - s) < len) {
+				t++;
+			}
+		}
+		/* update end of string with NUL */
+		*t = '\000';
+	}
+	return t - s;
+}
+
+/*
+ * console_setup(int baudrate)
+ *
+ * Set the pins and clocks to create a console that we can
+ * use for serial messages and getting text from the user.
+ */
+void console_setup(int baud)
+{
+	/* MUST enable the GPIO clock in ADDITION to the USART clock */
+	rcc_periph_clock_enable(RCC_GPIOA);
+
+	/* This example uses PD5 and PD6 for Tx and Rx respectively
+	 * but other pins are available for this role on USART1 (our chosen
+	 * USART) as well, such as PA2 and PA3. You can also split them
+	 * so PA2 for Tx, PD6 for Rx but you would have to enable both
+	 * the GPIOA and GPIOD clocks in that case
+	 */
+	gpio_mode_setup(GPIOA, GPIO_MODE_AF, GPIO_PUPD_NONE, GPIO9 | GPIO10);
+
+	/* Actual Alternate function number (in this case 7) is part
+	 * depenedent, CHECK THE DATA SHEET for the right number to
+	 * use.
+	 */
+	gpio_set_af(GPIOA, GPIO_AF7, GPIO9 | GPIO10);
+
+
+	/* This then enables the clock to the USART1 peripheral which is
+	 * attached inside the chip to the APB2 bus. Different peripherals
+	 * attach to different buses, and even some UARTS are attached to
+	 * APB1 and some to APB2, again the data sheet is useful here.
+	 */
+	rcc_periph_clock_enable(RCC_USART1);
+
+	/* Set up USART/UART parameters using the libopencm3 helper functions */
+	usart_set_baudrate(CONSOLE_UART, baud);
+	usart_set_databits(CONSOLE_UART, 8);
+	usart_set_stopbits(CONSOLE_UART, USART_STOPBITS_1);
+	usart_set_mode(CONSOLE_UART, USART_MODE_TX_RX);
+	usart_set_parity(CONSOLE_UART, USART_PARITY_NONE);
+	usart_set_flow_control(CONSOLE_UART, USART_FLOWCONTROL_NONE);
+	usart_enable(CONSOLE_UART);
+
+	/* Enable interrupts from the USART */
+	nvic_enable_irq(NVIC_USART1_IRQ);
+
+	/* Specifically enable receive interrupts */
+	usart_enable_rx_interrupt(CONSOLE_UART);
+}
+
+static ssize_t console_read(void *cookie, char *buf, size_t size)
+{
+	cookie = cookie;        /* -Wunused-parameter */
+	size_t i;
+	for (i = 0; i < size; i++) {
+		char c = console_getc(1);
+		buf[i] = c;
+		if (c == '\r') {
+			buf[i] = '\n';
+			i++;
+			break;
+		}
+	}
+	return i;
+}
+
+static ssize_t console_write(void *cookie, const char *buf, size_t size)
+{
+	cookie = cookie;        /* -Wunused-parameter */
+	size_t i;
+	for (i = 0; i < size; i++) {
+		char c = buf[i];
+		if (c == '\n') {
+			console_putc('\r');
+		}
+		console_putc(c);
+	}
+	return size;
+}
+
+void console_stdio_setup()
```

Comment:
I think it is a valid thing to not use the stdio functions. They end up adding a lot of bloat to your code, especially when you just try to print simple strings.

So I think @kbob approach is better than the _read _write one unless you really need/want the fancy printf capabilities. :)